File: backend-java/src/main/java/com/wealthtech/crm/modules/customer/service/ClientService.java
package com.wealthtech.crm.modules.customer.service;

import java.io.ByteArrayInputStream;
import java.time.LocalDate;
import java.util.List;

import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import com.wealthtech.crm.common.dto.DropdownOption;
import com.wealthtech.crm.modules.customer.dto.*;
import com.wealthtech.crm.modules.customer.entity.*;
import com.wealthtech.crm.modules.customer.enums.ClientStatus;
import com.wealthtech.crm.modules.customer.enums.KycStatus;
import com.wealthtech.crm.modules.customer.repository.*;
import com.wealthtech.crm.modules.usermanager.dto.CursorPaginatedResponse;
import com.wealthtech.crm.modules.usermanager.entity.User;
import com.wealthtech.crm.modules.usermanager.exception.ConflictException;
import com.wealthtech.crm.modules.usermanager.exception.NotFoundException;
import com.wealthtech.crm.modules.usermanager.repository.UserRepository;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
@RequiredArgsConstructor
public class ClientService {
    private final ClientRepository cRepo;
    private final ClientProfileRepository cpRepo;
    private final UserRepository userRepo;
    private final ClientExcelService excelService;

    // Mappers
    private ClientResponse mapToClientResponse(Client c) {
        DropdownOption<Long> rmDropdown = null;
        if(c.getRelationshipManager() != null) {
            String rmName = c.getRelationshipManager().getFirstName() + " " + c.getRelationshipManager().getLastName();
            rmDropdown = new DropdownOption<>(rmName, c.getRelationshipManager().getId());
        }

        String kycStatusStr = (c.getProfile() != null && c.getProfile().getKycStatus() != null)
                ? c.getProfile().getKycStatus().name() : KycStatus.PENDING.name();

        return new ClientResponse(c.getId(), c.getFirstName(),                                                                     
                    c.getLastName(),                                                                      
                    c.getFirstName() + " " + c.getLastName(),                                             
                    c.getEmail(),                                                                         
                    c.getPhone(),                                                                         
                    c.getPan(),                                                                           
                    c.getDateOfBirth(),                                                                   
                    c.getGender() != null ? c.getGender().name() : null,                                  
                    c.getStatus() != null ? c.getStatus().name() : null,                                  
                    kycStatusStr,                                                                     
                    rmDropdown,                                                                       
                    c.getSignUpDate(),                                                                    
                    c.getCreatedAt());
    }

    private ClientProfileResponse mapToProfileResponse(ClientProfile p) {
        return new ClientProfileResponse(                                                             
                    p.getId(),                                                                            
                    p.getClient() != null ? p.getClient().getId() : null,                                 
                    p.getKycStatus() != null ? p.getKycStatus().name() : null,                            
                    p.getClientStatus() != null ? p.getClientStatus().name() : null,                      
                    p.getAddressLine(),                                                                   
                    p.getCity(),                                                                          
                    p.getState(),                                                                         
                    p.getPincode(),                                                                       
                    p.getCountry(),                                                                       
                    p.getCreatedAt(),
                    p.getUpdatedAt()
            );
    }

    // Read operations
    public ClientResponse getClient(Long id) {
        Client client = cRepo.findByIdWithRelations(id)
                    .orElseThrow(() -> new NotFoundException("Client not found with id: " + id));
        return mapToClientResponse(client);
    }

    public ClientProfileResponse getClientProfile(Long clientId) {
        ClientProfile profile = cpRepo.findByClientId(clientId) .orElseThrow(() -> new NotFoundException("Client profile not found client id: " + clientId));
        return mapToProfileResponse(profile);
    }

    public CursorPaginatedResponse<ClientResponse> getClientList(
        ClientStatus status,
        Long rmId,
        String search,
        Long cursor,
        int pageSize
    ) {
        long effectiveCursor = (cursor != null) ? cursor : 0L;
        Pageable pageable = PageRequest.of(0, pageSize+1);

        List<Client> clients = cRepo.findWithFilters(status, rmId, search, effectiveCursor, pageable);
        
        boolean hasNext = clients.size() > pageSize;
        List<Client> pageResults = hasNext ? clients.subList(0, pageSize) : clients;

        String nextCursor = null;
        if(hasNext && !pageResults.isEmpty()) {
            nextCursor = String.valueOf(pageResults.get(pageResults.size() - 1).getId());
        }

        long totalSize = cRepo.countWithFilters(status, rmId, search);
        
        List<ClientResponse> responseList = pageResults.stream().map(this::mapToClientResponse).toList();

        return new CursorPaginatedResponse<>(responseList, nextCursor, null, null, null, totalSize);
    }

    // Create and  update operations
    @Transactional 
    public ClientResponse createClient(CreateClientRequest request, Long currentUserId) {
        if(cRepo.existsByEmail(request.email())) {
            throw new ConflictException("Client with email already exists: " + request.email());
        }

        if(request.phone() != null && !request.phone().isBlank() && cRepo.existsByPhone(request.phone())) {
            throw new ConflictException("Client with phone already exists: " + request.phone());
        }
        
        if(request.pan() != null && !request.pan().isBlank() && cRepo.existsByPan(request.pan())) {
            throw new ConflictException("Client with pan already exists: "+ request.pan());
        }

        // Resolve relationship manager
        User rm = null;
        if(request.relationshipManagerId() != null) {
            rm = userRepo.findById(request.relationshipManagerId())
                    .orElseThrow(() -> new NotFoundException("Relationship Manager not found: " + request.relationshipManagerId()));
        }
        else if(currentUserId != null) {
            // Automatically assign the logged in employee as the RM
            rm = userRepo.findById(currentUserId).orElse(null);
        }

        // Build client entity
        Client client = Client.builder()
                        .firstName(request.firstName())
                        .lastName(request.lastName())
                        .email(request.email())
                        .phone(request.phone())
                        .pan(request.pan())
                        .dateOfBirth(request.dateOfBirth())
                        .gender(request.gender())
                        .status(ClientStatus.ONBOARDING)
                        .relationshipManager(rm)
                        .signUpDate(LocalDate.now())
                        .createdBy(currentUserId)
                        .build();
        
        // Build lean ClientProfile
        ClientProfile profile = ClientProfile.builder()
                    .client(client)
                    .kycStatus(KycStatus.PENDING)
                    .clientStatus(ClientStatus.ONBOARDING)
                    .addressLine(request.addressLine())
                    .city(request.city())
                    .state(request.state())
                    .pincode(request.pincode())
                    .country(request.country() != null ? request.country() : "India")
                    .build();
        client.setProfile(profile);
        Client savedClient = cRepo.save(client);
        return mapToClientResponse(savedClient);
    }

    @Transactional
    public ClientResponse updateClientStatus(Long id, UpdateClientStatusRequest request) {
        Client client = cRepo.findByIdWithRelations(id)
                    .orElseThrow(() -> new NotFoundException("Client not found with id: " + id));
        
        client.setStatus(request.status());
        Client updatedClient = cRepo.save(client);
        return mapToClientResponse(updatedClient);
    }

    @Transactional
    public ClientResponse updateClientRm(Long id, UpdateClientRmRequest request) {
        Client client = cRepo.findByIdWithRelations(id)
                .orElseThrow(() -> new NotFoundException("Client not found with id: " + id));
        
        User rm = userRepo.findById(request.rmId())
                    .orElseThrow(() -> new NotFoundException("Relationship manager not found with id: " + request.rmId()));
        
        client.setRelationshipManager(rm);
        Client updatedClient = cRepo.save(client);
        return mapToClientResponse(updatedClient);
    }

    @Transactional                                                                                                                    
    public int bulkReassignRm(BulkReassignRmRequest request) {
            User newRm = userRepo.findById(request.newRmId())
                    .orElseThrow(() -> new NotFoundException("Relationship Manager not found with id: " + request.newRmId()));            
  
            return cRepo.bulkReassignRm(request.clientIds(), newRm);
        }
    
    @Transactional
    public ClientProfileResponse verifyKyc(Long clientId, VerifyKycRequest request) {
        Client client = cRepo.findByIdWithRelations(clientId).orElseThrow(() -> new NotFoundException("Client not found with id: " + clientId));

        ClientProfile profile = client.getProfile();
        if(profile == null) {
            throw new NotFoundException("Client profile not found for client id: " + clientId);
        }

        profile.setKycStatus(request.kycStatus());

        if(request.kycStatus() == KycStatus.VERIFIED) {
            client.setStatus(ClientStatus.ACTIVE);
            cRepo.save(client);
        }

        ClientProfile updatedProfile = cpRepo.save(profile);
        return mapToProfileResponse(updatedProfile);
    }

    @Async
    public void processBulkUploadAsync(byte[] fileBytes, Long currentUserId) {
        log.info("Starting async bulk client upload for userId: {}", currentUserId);
        try (ByteArrayInputStream bais = new ByteArrayInputStream(fileBytes)) {
            List<CreateClientRequest> requests = excelService.parseClientExcel(bais);
            log.info("Parsed {} client records from uploaded Excel file", requests.size());

            int savedCount = 0;
            int skippedCount = 0;

            for (CreateClientRequest req : requests) {
                // Skip duplicates
                if (cRepo.existsByEmail(req.email()) ||
                    (req.phone() != null && !req.phone().isBlank() && cRepo.existsByPhone(req.phone())) ||
                    (req.pan() != null && !req.pan().isBlank() && cRepo.existsByPan(req.pan()))) {
                    log.warn("Skipping duplicate client: email={}, pan={}", req.email(), req.pan());
                    skippedCount++;
                    continue;
                }

                createClient(req, currentUserId);
                savedCount++;
            }

            log.info("Async bulk client upload finished. Successfully saved: {}, Skipped duplicates: {}", savedCount, skippedCount);
        } catch (Exception e) {
            log.error("Failed to process bulk client upload asynchronously", e);
        }
    }
}

File: backend-nodejs/src/modules/customer/services/clientService.ts
import { FilterQuery, Types } from 'mongoose';
import { Client, IClient } from '../models/Client';
import { ClientProfile, IClientProfile } from '../models/ClientProfile';
import { User } from '../../usermanager/models/User';
import {
  CreateClientDto,
  UpdateClientStatusDto,
  UpdateClientRmDto,
  VerifyKycDto,
  BulkReassignRmDto,
  ClientResponseDto,
  ClientProfileResponseDto,
} from '../dto/clientDto';
import { clientExcelService } from './clientExcelService';
import { DropdownOption } from '../../../common/types/dropdown';
import { ClientStatus, KycStatus } from '../enums/clientEnums';
import { CursorPaginationResponse } from '../../../common/types/pagination';
import { logger } from '../../../common/utils/logger';
import { AppError } from '../../../common/utils/AppError';

interface PopulatedRm {
  _id: Types.ObjectId;
  fullName?: string;
  email: string;
}

type PopulatedClient = Omit<IClient, 'relationshipManager'> & {
  relationshipManager?: Types.ObjectId | PopulatedRm | null;
};

export class ClientService {
  // Helper to format a populated Client document into ClientResponseDto
  private mapToClientResponse(client: PopulatedClient, profile?: IClientProfile | null): ClientResponseDto {
    let rmDropdown: DropdownOption<string> | null = null;

    if (
      client.relationshipManager &&
      typeof client.relationshipManager === 'object' &&
      '_id' in client.relationshipManager
    ) {
      const rm = client.relationshipManager as PopulatedRm;
      rmDropdown = {
        display_name: rm.fullName || rm.email,
        value: rm._id.toString(),
      };
    }

    const kycStatus = profile?.kycStatus || KycStatus.PENDING;

    return {
      id: client._id.toString(),
      firstName: client.firstName,
      lastName: client.lastName,
      fullName: client.fullName || `${client.firstName} ${client.lastName}`.trim(),
      email: client.email,
      phone: client.phone,
      pan: client.pan,
      dateOfBirth: client.dateOfBirth,
      gender: client.gender,
      status: client.status,
      kycStatus,
      relationshipManager: rmDropdown,
      signUpDate: client.signUpDate,
      createdAt: client.createdAt,
    };
  }

  // Format ClientProfile document into ClientProfileResponseDto
  private mapToProfileResponse(profile: IClientProfile): ClientProfileResponseDto {
    return {
      id: profile._id.toString(),
      clientId: profile.client.toString(),
      kycStatus: profile.kycStatus,
      clientStatus: profile.clientStatus,
      addressLine: profile.addressLine,
      city: profile.city,
      state: profile.state,
      pincode: profile.pincode,
      country: profile.country,
      createdAt: profile.createdAt,
      updatedAt: profile.updatedAt,
    };
  }

  // List clients with cursor pagination and multi-field search/filtering
  async getClients(params: {
    status?: ClientStatus;
    rmId?: string;
    search?: string;
    cursor?: string;
    pageSize?: number;
  }): Promise<CursorPaginationResponse<ClientResponseDto>> {
    const limit = Math.min(params.pageSize || 50, 1500);
    const filter: FilterQuery<IClient> = {};

    if (params.status) {
      filter.status = params.status;
    }

    if (params.rmId) {
      filter.relationshipManager = new Types.ObjectId(params.rmId);
    }

    if (params.search) {
      const searchRegex = { $regex: String(params.search), $options: 'i' };
      filter.$or = [
        { firstName: searchRegex },
        { lastName: searchRegex },
        { email: searchRegex },
        { phone: searchRegex },
        { pan: searchRegex },
      ];
    }

    if (params.cursor) {
      filter._id = { $gt: new Types.ObjectId(params.cursor) };
    }

    const [clients, totalCount] = await Promise.all([
      Client.find(filter)
        .populate('relationshipManager', 'fullName email')
        .sort({ _id: 1 })
        .limit(limit + 1),
      Client.countDocuments(params.search || params.status || params.rmId ? filter : {}),
    ]);

    const hasNextPage = clients.length > limit;
    const results = hasNextPage ? clients.slice(0, limit) : clients;
    const nextCursor = hasNextPage ? results[results.length - 1]._id.toString() : null;

    // Fetch corresponding profiles in parallel to attach kycStatus
    const clientIds = results.map((c) => c._id);
    const profiles = await ClientProfile.find({ client: { $in: clientIds } });
    const profileMap = new Map(profiles.map((p) => [p.client.toString(), p]));

    const responseResults = results.map((c) =>
      this.mapToClientResponse(c as unknown as PopulatedClient, profileMap.get(c._id.toString()))
    );

    return {
      results: responseResults,
      next: nextCursor,
      previous: null,
      pageNumber: null,
      totalPages: null,
      totalSize: totalCount,
    };
  }

  // Get single client by ID
  async getClientById(id: string): Promise<ClientResponseDto> {
    const client = await Client.findById(id).populate('relationshipManager', 'fullName email');
    if (!client) {
      logger.warn({ clientId: id }, 'Client lookup failed: Client not found');
      throw new AppError('Client not found', 404);
    }

    const profile = await ClientProfile.findOne({ client: client._id });
    return this.mapToClientResponse(client as unknown as PopulatedClient, profile);
  }

  // Get client KYC profile
  async getClientProfile(clientId: string): Promise<ClientProfileResponseDto> {
    const profile = await ClientProfile.findOne({ client: clientId });
    if (!profile) {
      logger.warn({ clientId }, 'Client profile lookup failed: Profile not found');
      throw new AppError('Client profile not found', 404);
    }

    return this.mapToProfileResponse(profile);
  }

  // Create client and associated profile
  async createClient(data: CreateClientDto, currentUserId?: string): Promise<ClientResponseDto> {
    const { firstName, lastName, email, phone, pan } = data;

    // Check unique email
    const existingEmail = await Client.findOne({ email: email.toLowerCase().trim() });
    if (existingEmail) {
      logger.warn({ email }, 'Client creation failed: Email already exists');
      throw new AppError(`Client with email '${email}' already exists`, 409);
    }

    // Check unique phone
    const existingPhone = await Client.findOne({ phone: phone.trim() });
    if (existingPhone) {
      logger.warn({ phone }, 'Client creation failed: Phone already exists');
      throw new AppError(`Client with phone '${phone}' already exists`, 409);
    }

    // Check unique PAN (if provided)
    if (pan && pan.trim()) {
      const existingPan = await Client.findOne({ pan: pan.trim().toUpperCase() });
      if (existingPan) {
        logger.warn({ pan }, 'Client creation failed: PAN already exists');
        throw new AppError(`Client with PAN '${pan}' already exists`, 409);
      }
    }

    // Resolve RM - explicit RM assignment or auto-assign current user
    let rmId: Types.ObjectId | null = null;
    if (data.relationshipManagerId) {
      const rmUser = await User.findById(data.relationshipManagerId);
      if (!rmUser) {
        logger.warn({ rmId: data.relationshipManagerId }, 'Client creation failed: RM user not found');
        throw new AppError('Specified Relationship Manager does not exist', 404);
      }
      rmId = rmUser._id;
    } else if (currentUserId) {
      rmId = new Types.ObjectId(currentUserId);
    }

    // Create Client document
    const client = await Client.create({
      firstName: firstName.trim(),
      lastName: lastName.trim(),
      email: email.toLowerCase().trim(),
      phone: phone.trim(),
      pan: pan ? pan.trim().toUpperCase() : undefined,
      dateOfBirth: data.dateOfBirth,
      gender: data.gender,
      status: ClientStatus.ONBOARDING,
      relationshipManager: rmId,
      signUpDate: new Date(),
      createdBy: currentUserId ? new Types.ObjectId(currentUserId) : undefined,
    });

    // Create corresponding ClientProfile document
    const profile = await ClientProfile.create({
      client: client._id,
      kycStatus: KycStatus.PENDING,
      clientStatus: ClientStatus.ONBOARDING,
      addressLine: data.addressLine,
      city: data.city,
      state: data.state,
      pincode: data.pincode,
      country: data.country || 'India',
    });

    const populatedClient = await Client.findById(client._id).populate(
      'relationshipManager',
      'fullName email'
    );

    logger.info({ clientId: client._id, email: client.email }, 'Client and profile created successfully');
    return this.mapToClientResponse(populatedClient as unknown as PopulatedClient, profile);
  }

  // Update client operational status
  async updateClientStatus(id: string, data: UpdateClientStatusDto): Promise<ClientResponseDto> {
    const client = await Client.findById(id).populate('relationshipManager', 'fullName email');
    if (!client) {
      logger.warn({ clientId: id }, 'Update status failed: Client not found');
      throw new AppError('Client not found', 404);
    }

    client.status = data.status;
    await client.save();

    // Synchronize status in ClientProfile
    const profile = await ClientProfile.findOne({ client: client._id });
    if (profile) {
      profile.clientStatus = data.status;
      await profile.save();
    }

    logger.info({ clientId: id, newStatus: data.status }, 'Client status updated successfully');
    return this.mapToClientResponse(client as unknown as PopulatedClient, profile);
  }

  // Update client Relationship Manager
  async updateClientRm(id: string, data: UpdateClientRmDto): Promise<ClientResponseDto> {
    const client = await Client.findById(id);
    if (!client) {
      logger.warn({ clientId: id }, 'Update RM failed: Client not found');
      throw new AppError('Client not found', 404);
    }

    const rmUser = await User.findById(data.rmId);
    if (!rmUser) {
      logger.warn({ rmId: data.rmId }, 'Update RM failed: RM user not found');
      throw new AppError('Relationship Manager not found', 404);
    }

    client.relationshipManager = rmUser._id;
    await client.save();

    const populatedClient = await Client.findById(client._id).populate(
      'relationshipManager',
      'fullName email'
    );
    const profile = await ClientProfile.findOne({ client: client._id });

    logger.info({ clientId: id, rmId: data.rmId }, 'Client RM reassigned successfully');
    return this.mapToClientResponse(populatedClient as unknown as PopulatedClient, profile);
  }

  // Verify KYC (approving KYC activates client)
  async verifyKyc(clientId: string, data: VerifyKycDto): Promise<ClientProfileResponseDto> {
    const client = await Client.findById(clientId);
    if (!client) {
      logger.warn({ clientId }, 'Verify KYC failed: Client not found');
      throw new AppError('Client not found', 404);
    }

    const profile = await ClientProfile.findOne({ client: clientId });
    if (!profile) {
      logger.warn({ clientId }, 'Verify KYC failed: Profile not found');
      throw new AppError('Client profile not found', 404);
    }

    profile.kycStatus = data.kycStatus;

    // Automatic Lifecycle Transition: Verified KYC activates the client
    if (data.kycStatus === KycStatus.VERIFIED) {
      client.status = ClientStatus.ACTIVE;
      profile.clientStatus = ClientStatus.ACTIVE;
      await client.save();
    }

    await profile.save();
    logger.info({ clientId, kycStatus: data.kycStatus }, 'Client KYC verified and status updated');
    return this.mapToProfileResponse(profile);
  }

  // Bulk reassign Relationship Manager
  async bulkReassignRm(data: BulkReassignRmDto): Promise<{ message: string; count: number }> {
    const rmUser = await User.findById(data.newRmId);
    if (!rmUser) {
      logger.warn({ rmId: data.newRmId }, 'Bulk reassign failed: New RM user not found');
      throw new AppError('Relationship Manager not found', 404);
    }

    const clientObjectIds = data.clientIds.map((id: string) => new Types.ObjectId(id));
    const result = await Client.updateMany(
      { _id: { $in: clientObjectIds } },
      { $set: { relationshipManager: rmUser._id } }
    );

    logger.info(
      { newRmId: data.newRmId, modifiedCount: result.modifiedCount },
      'Bulk reassigned relationship manager'
    );

    return {
      message: 'Successfully reassigned relationship manager',
      count: result.modifiedCount,
    };
  }

  // Process bulk upload asynchronously in background
  async processBulkUploadAsync(buffer: Buffer, currentUserId?: string): Promise<void> {
    logger.info({ currentUserId }, 'Starting background bulk client upload ingestion');
    try {
      const requests = await clientExcelService.parseClientExcel(buffer);
      logger.info({ totalRecords: requests.length }, 'Parsed records from uploaded Excel');

      let savedCount = 0;
      let skippedCount = 0;

      for (const req of requests) {
        // Skip duplicate records
        const duplicate = await Client.findOne({
          $or: [
            { email: req.email },
            { phone: req.phone },
            ...(req.pan ? [{ pan: req.pan }] : []),
          ],
        });

        if (duplicate) {
          logger.warn(
            { email: req.email, phone: req.phone, pan: req.pan },
            'Skipping duplicate client in bulk upload'
          );
          skippedCount++;
          continue;
        }

        try {
          await this.createClient(req, currentUserId);
          savedCount++;
        } catch (err: unknown) {
          const errMsg = err instanceof Error ? err.message : 'Unknown error';
          logger.error({ email: req.email, err: errMsg }, 'Failed to save client in bulk upload');
          skippedCount++;
        }
      }

      logger.info(
        { savedCount, skippedCount },
        'Background bulk client upload ingestion finished'
      );
    } catch (error: unknown) {
      const errMsg = error instanceof Error ? error.message : 'Unknown error';
      logger.error({ err: errMsg }, 'Failed to process bulk client upload');
    }
  }
}

export const clientService = new ClientService();